### PR TITLE
Speed up the override registration phase

### PR DESCRIPTION
**Pull request by robdockins**
by prefiltering possible overrides first.

This significantly reduces the amount of startup time that is consumed
by installing overrides, especially C++ overrides that perform name demangling.

Each override that wants to be registered now provides a "matcher" which examines
the name of a symbol for an approximate test as to weather this override applies.
These can either be exact matches, prefix matches, or substring matches.  Any
override that passes the prefilter test is considered in order, as before, to
determine more precicely if it applies.

On one modestly-sized example, this reduced overall verification time from
105 seconds to 32 seconds.  The algorithm used here for pre-filtering is
quite naive, so additional improvements are perhaps possible.  However,
at the momemnt, startup time is dominated by parsing the bitcode file itself.

Partially addresses #155

### CODE REVIEW

File: crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Common.hs [hunk 1]
```diff
@@ -99,8 +105,20 @@ data SomeLLVMOverride p sym arch =
 llvmSizeT :: HasPtrWidth wptr => L.Type
 llvmSizeT = L.PrimType $ L.Integer $ fromIntegral $ natValue $ PtrWidth
 
+data OverrideTemplate p sym arch rtp l a =
+  OverrideTemplate
+  { overrideTemplateMatcher :: TemplateMatcher
+  , overrideTemplateAction :: RegOverrideM p sym arch rtp l a ()
+  }
+
```
Comment (siddharthist) [inline crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Common.hs]:
```suggestion

-- | This type controls whether an override is installed for a given name found in a module.
--  See 'filterTemplates'.
```

File: crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Libcxx.hs [hunk 2]
```diff
@@ -70,33 +70,41 @@ import           Lang.Crucible.LLVM.Intrinsics.Common
...
+           Nothing -> empty
+           Just (SomeLLVMOverride override) -> register_llvm_override override
+
 
 -- type CPPOverride p sym arch args ret =
 --   L.Declare -> LLVMContext arch -> Maybe (LLVMOverride p sym arch args ret)
 
-type SomeCPPOverride p sym arch =
-  L.Declare -> LLVMContext arch -> Maybe (SomeLLVMOverride p sym arch)
+data SomeCPPOverride p sym arch =
```
Comment (siddharthist) [inline crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Libcxx.hs]:
```suggestion
-- | We can only tell whether we should install a C++ override after demangling
--  the function name, which is expensive. As a first approximation, we ask whether
--  the function's name contains a few substrings, in order. 
data SomeCPPOverride p sym arch =
```

**Comment by siddharthist** [comment]
Huh, the CI failure looks unrelated:
```haskell
Configuring boomerang-1.3.1...
Preprocessing library for boomerang-1.3.1..
Building library for boomerang-1.3.1..
[1 of 9] Compiling Text.Boomerang.HStack ( Text/Boomerang/HStack.hs, dist/build/Text/Boomerang/HStack.o )
[2 of 9] Compiling Text.Boomerang.Pos ( Text/Boomerang/Pos.hs, dist/build/Text/Boomerang/Pos.o )
[3 of 9] Compiling Text.Boomerang.Prim ( Text/Boomerang/Prim.hs, dist/build/Text/Boomerang/Prim.o )
Text/Boomerang/Prim.hs:57:10: error:
    • No instance for (Applicative (Parser e tok))
        arising from the superclasses of an instance declaration
    • In the instance declaration for ‘Monad (Parser e tok)’
   |
57 | instance Monad (Parser e tok) where
   |          ^^^^^^^^^^^^^^^^^^^^
```

**Comment by robdockins** [comment]
That CI failure is really strange... it's trying to build a really old version of boomerang.

**Pull request merged**